mongoengine/connection.py

```diff
@@ -177,10 +192,12 @@ def get_db(alias=DEFAULT_CONNECTION_NAME, reconnect=False):
         conn_settings = _connection_settings[alias]
         db = conn[conn_settings['name']]
         # Authenticate if necessary
-        if conn_settings['username'] and conn_settings['password']:
+        if (conn_settings['authentication_mechanism'] == 'MONGODB-X509' and conn_settings['username']) or\
+                (conn_settings['username'] and conn_settings['password']):
             db.authenticate(conn_settings['username'],
                             conn_settings['password'],
-                            source=conn_settings['authentication_source'])
+                            source=conn_settings['authentication_source'],
+                            mechanism=conn_settings['authentication_mechanism'])
```

Comment:
Given my comment above about PyMongo's default behavior, I think that we could default `authentication_mechanism` to None and then we could modify this code to do:

```
# include the authentication mechanism in kwargs if specified
auth_kwargs = {'source': conn_settings['authentication_source']}
if conn_settings['authenticaltion_mechanism'] is not None:
    auth_kwargs['mechanism'] = conn_settings['authenticaltion_mechanism']
db.authenticate(conn_settings['username'], conn_settings['password'], **kwargs)
```